api/explorer.go

```diff
@@ -1,62 +1,334 @@
-func (srv *Server) explorerGetHashHandler(w http.ResponseWriter, req *http.Request) {
-	// Extract the hash from the request
-	var data []byte
-	_, err := fmt.Sscanf(req.FormValue("hash"), "%x", &data)
-	if err != nil {
-		writeError(w, "Malformed or no hash provided: "+err.Error(), http.StatusBadRequest)
+	// Try the hash as an unlock hash.
+	txids := srv.explorer.UnlockHash(types.UnlockHash(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
+					ID:       block.ID(),
+					Height:   height,
+					RawBlock: block,
+				})
+			}
+			for _, t := range block.Transactions {
+				if t.ID() == txid {
+					txns = append(txns, ExplorerTransaction{
+						ID: txid,
+						Parent: ExplorerBlock{
+							ID:       block.ID(),
+							Height:   height,
+							RawBlock: block,
+						},
+						RawTransaction: t,
+					})
+				}
+			}
+		}
+		writeJSON(w, ExplorerHashGET{
+			HashType:     "unlockhash",
+			Blocks:       blocks,
+			Transactions: txns,
+		})
 		return
 	}
 
-	// returnData will be a generic interface. The json encoder
-	// should still work though
-	returnData, err := srv.exp.GetHashInfo(data)
-	if err != nil {
-		writeError(w, err.Error(), http.StatusInternalServerError)
+	// Try the hash as a siacoin output id.
+	txids = srv.explorer.SiacoinOutputID(types.SiacoinOutputID(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
+					ID:       block.ID(),
+					Height:   height,
+					RawBlock: block,
+				})
+			}
+			for _, t := range block.Transactions {
+				if t.ID() == txid {
+					txns = append(txns, ExplorerTransaction{
+						ID: txid,
+						Parent: ExplorerBlock{
+							ID:       block.ID(),
+							Height:   height,
+							RawBlock: block,
+						},
+						RawTransaction: t,
+					})
+				}
+			}
+		}
+		writeJSON(w, ExplorerHashGET{
+			HashType:     "siacoinoutputid",
+			Blocks:       blocks,
+			Transactions: txns,
+		})
+		return
+	}
+
+	// Try the hash as a file contract id.
+	txids = srv.explorer.FileContractID(types.FileContractID(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
```

Comment:
why does all this logic happen in the API package instead of inside the explorer?